src/components/Datagrid/useDatagrid.tsx

```diff
@@ -0,0 +1,155 @@
+import { useCallback, useEffect, useRef, useState } from "react";
+import { Matrix } from "react-spreadsheet";
+
+import { MoneyCellEdit, MoneyCellView } from "./MoneyCell";
+import { MoneyToggleCellView } from "./MoneyToggleCell";
+import { NumberCellEdit, NumberCellView } from "./NumberCell";
+import { AvailableColumn, ColumnState, DatagridCell } from "./types";
+
+function getId(row: DatagridCell[]): string {
+  return row[0].id;
+}
+
+export interface UseDatagridOpts<T> {
+  availableColumns: AvailableColumn[];
+  data: T[];
+  getData: (row: T, column: string) => DatagridCell;
+  onChange: (data: Matrix<DatagridCell>) => void;
+}
+
+function getViewer(type: DatagridCell["type"]) {
+  switch (type) {
+    case "money":
+      return MoneyCellView;
+    case "moneyToggle":
+      return MoneyToggleCellView;
+    case "number":
+      return NumberCellView;
+    default:
+      return undefined;
+  }
+}
+
+function getEditor(type: DatagridCell["type"]) {
+  switch (type) {
+    case "money":
+    case "moneyToggle":
+      return MoneyCellEdit;
+    case "number":
+      return NumberCellEdit;
+    default:
+      return undefined;
+  }
+}
+
+function useDatagrid<T extends { id: string }>({
+  availableColumns,
+  data: initial,
+  getData,
+  onChange: onChangeBase
+}: UseDatagridOpts<T>) {
+  const [columns, setColumns] = useState<ColumnState[]>(
+    availableColumns.map(c => ({
+      ...c,
+      width: 200
+    }))
+  );
+  const [rows, setRows] = useState<string[]>(initial.map(d => d.id));
+
+  // May contain formulas
+  const [data, setData] = useState<Matrix<DatagridCell>>([]);
+  const dataRef = useRef<Matrix<DatagridCell>>(data);
+  const updateRef = (d: Matrix<DatagridCell>) => {
+    dataRef.current = d;
+  };
```

Comment:
I really don't get why we're using `data` and `dataRef` at the same time. Is it because of that infinite loop?
I would really appreciate a comment why that is to avoid problems if we need to update this hook in the future